hw/ip/otbn/util/otbn_sim_test.py

```diff
@@ -63,49 +81,89 @@ def main() -> int:
+            args.elf
+        ]
+        # Run the simulation and produce a register and dmem dump.
+        sim_proc = subprocess.run(
+            cmd, check=True, stdout=subprocess.PIPE, universal_newlines=True
+        )
+
+        if args.expected_dmem is not None:
+            dmem_file.seek(0)
+            actual_dmem = parse_actual_dmem(dmem_file.read())
+            expected_dmem = parse_dmem_exp(args.expected_dmem.read())
+
+    if args.expected is not None:
+        expected_regs = parse_reg_dump(args.expected.read())
+        actual_regs = parse_reg_dump(sim_proc.stdout)
 
     # Special handling for the ERR_BITS register.
-    expected_err = expected_regs.get(ERR_BITS, 0)
-    actual_err = actual_regs[ERR_BITS]
-    insn_cnt = actual_regs[INSN_CNT]
-    stop_pc = actual_regs[STOP_PC]
-    if expected_err == 0 and actual_err != 0:
-        # Test is expected to have no errors, but an error occurred. In this
-        # case, give a special error message and exit rather than print all the
-        # mismatched registers.
-        if actual_err != 0:
-            err_names = ', '.join(get_err_names(actual_err))
-            result.err(f'OTBN encountered an unexpected error: {err_names}.\n'
-                       f'  {ERR_BITS}\t= {actual_err:#010x}\n'
-                       f'  {INSN_CNT}\t= {insn_cnt:#010x}\n'
-                       f'  {STOP_PC}\t= {stop_pc:#010x}')
-    else:
-        for reg, expected_value in expected_regs.items():
-            actual_value = actual_regs.get(reg, None)
-            if actual_value != expected_value:
-                if reg.startswith('w'):
-                    expected_str = f'{expected_value:#066x}'
-                    actual_str = f'{actual_value:#066x}'
-                else:
-                    expected_str = f'{expected_value:#010x}'
-                    actual_str = f'{actual_value:#010x}'
-                result.err(f'Mismatch for register {reg}:\n'
-                           f'  Expected: {expected_str}\n'
-                           f'  Actual:   {actual_str}')
+    if args.expected is not None:
+        expected_err = expected_regs.get(ERR_BITS, 0)
+        actual_err = actual_regs[ERR_BITS]
+        insn_cnt = actual_regs[INSN_CNT]
+        stop_pc = actual_regs[STOP_PC]
+        if expected_err == 0 and actual_err != 0:
+            # Test is expected to have no errors, but an error occurred. In this
+            # case, give a special error message and exit rather than print all the
+            # mismatched registers.
+            if actual_err != 0:
+                err_names = ", ".join(get_err_names(actual_err))
+                result.err(
+                    f"OTBN encountered an unexpected error: {err_names}.\n"
+                    f"  {ERR_BITS}\t= {actual_err:#010x}\n"
+                    f"  {INSN_CNT}\t= {insn_cnt:#010x}\n"
+                    f"  {STOP_PC}\t= {stop_pc:#010x}"
+                )
```

Comment:
It might make sense to move this logic out of the if statement, and just use an expected error value of 0 (no errors) if no expected-register file is provided. Right now, I think if there are no expected register values provided and the program encounters an error, this rule will only check the values in DMEM and won't say that there was an error at all. Like for registers, I think it also makes sense in that case not to print the mismatched DMEM values, since the end of the test was not reached.